subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -98,24 +104,29 @@ public DependentsSet getRelevantDependents(String className, IntSet constants) {
         if (fullRebuildCause != null) {
             return DependentsSet.dependencyToAll(fullRebuildCause);
         }
+
         DependentsSet deps = getDependents(className);
         if (deps.isDependencyToAll()) {
             return deps;
         }
-        if (!constants.isEmpty()) {
+        if (!constants.isEmpty() && !compilerApiData.isAvailable()) {
             return DependentsSet.dependencyToAll();
         }
         Set<String> classesDependingOnAllOthers = annotationProcessingData.participatesInClassGeneration(className) ? annotationProcessingData.getGeneratedTypesDependingOnAllOthers() : Collections.emptySet();
         Set<GeneratedResource> resourcesDependingOnAllOthers = annotationProcessingData.participatesInResourceGeneration(className) ? annotationProcessingData.getGeneratedResourcesDependingOnAllOthers() : Collections.emptySet();
-        if (!deps.hasDependentClasses() && classesDependingOnAllOthers.isEmpty() && resourcesDependingOnAllOthers.isEmpty()) {
+        Set<String> constantDependants = getConstantDependents(className);
+        if (!deps.hasDependentClasses() && classesDependingOnAllOthers.isEmpty() && resourcesDependingOnAllOthers.isEmpty() && constantDependants.isEmpty()) {
             return deps;
         }
 
+        Set<String> dependents = new HashSet<>(deps.getAccessibleDependentClasses());
+
         Set<String> privateResultClasses = new HashSet<>();
         Set<String> accessibleResultClasses = new HashSet<>();
         Set<GeneratedResource> resultResources = new HashSet<>(resourcesDependingOnAllOthers);
-        processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, deps.getPrivateDependentClasses(), deps.getAccessibleDependentClasses());
+        processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, deps.getPrivateDependentClasses(), dependents);
         processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, Collections.emptySet(), classesDependingOnAllOthers);
+        processDependentClasses(new HashSet<>(), privateResultClasses, accessibleResultClasses, resultResources, Collections.emptySet(), constantDependants);
```

Comment:
Is here enough, that I check only constants dependents recursively or should I get all transient dependents of constant dependents?
So first option is:
- A has CONSTANT1
- B depends on CONSTANT1 and has CONSTANT2
Find constant dependents of A (this is B) and then recursively constant dependents of B 

Second one is:
- A has CONSTANT1
- B depends on CONSTANT1 and has CONSTANT2
Find constant dependents of A (this is B) and then recursively ALL dependents of B 

Currently the second one is implemented, but first solution would "save" few classes to compile in some cases.

Reply:
obsolete